core-bundle/src/EventListener/PreviewToolbarListener.php

```diff
@@ -0,0 +1,161 @@
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpKernel\Event\ResponseEvent;
+use Symfony\Component\Routing\RouterInterface;
+use Symfony\Component\Security\Core\Authentication\Token\Storage\TokenStorageInterface;
+use Symfony\Component\Security\Core\Authorization\AuthorizationChecker;
+use Symfony\Component\Security\Csrf\CsrfTokenManagerInterface;
+use Twig\Environment as TwigEnvironment;
+use Twig\Error\Error as TwigError;
+
+/**
+ * PreviewToolbarListener injects the back end preview switch toolbar.
+ *
+ * The onKernelResponse method must be connected to the kernel.response event.
+ *
+ * The toolbar is only injected on well-formed HTML (with a proper <body> tag).
+ * This means that the WDT is never included in sub-requests or ESI requests.
+ */
+final class PreviewToolbarListener
+{
+
+    private $scopeMatcher;
+
+    private $previewScript;
+
+    private $twig;
+
+    private $authorizationChecker;
+
+    private $tokenStorage;
+
+    private $tokenChecker;
+
+    private $router;
+
+    private $tokenManager;
+
+    private $csrfTokenName;
+
+    public function __construct(
+        string $previewScript,
+        ScopeMatcher $scopeMatcher,
+        TwigEnvironment $twig,
+        AuthorizationChecker $authorizationChecker,
+        TokenStorageInterface $tokenStorage,
+        TokenChecker $tokenChecker,
+        RouterInterface $router,
+        CsrfTokenManagerInterface $tokenManager,
+        string $csrfTokenName
+    ) {
+        $this->previewScript        = $previewScript;
+        $this->scopeMatcher         = $scopeMatcher;
+        $this->twig                 = $twig;
+        $this->authorizationChecker = $authorizationChecker;
+        $this->tokenStorage         = $tokenStorage;
+        $this->tokenChecker         = $tokenChecker;
+        $this->router               = $router;
+        $this->tokenManager         = $tokenManager;
+        $this->csrfTokenName        = $csrfTokenName;
+    }
+
+    public function onKernelResponse(ResponseEvent $event): void
+    {
+        if (!$this->scopeMatcher->isContaoMasterRequest($event)) {
+            return;
+        }
+
+        $request  = $event->getRequest();
+        $response = $event->getResponse();
+
+        if ($request->getScriptName() !== $this->previewScript) {
+            return;
+        }
+
+        // Ignore redirects and errors
+        if (200 !== $response->getStatusCode()) {
+            return;
+        }
+
+        // Do not capture redirects or modify XML HTTP Requests
+        if ($request->isXmlHttpRequest()) {
+            return;
+        }
+
+        if (null === $user = BackendUser::getInstance()) {
+            return;
+        }
+
+        // Only inject toolbar on html responses
+        if ('html' !== $request->getRequestFormat()
+            || false === strpos((string) $response->headers->get('Content-Type'), 'html')
```

Comment:
We check for `text/html` everywhere else :)

Reply:
So can be removed safely?

Reply:
No it should be adjusted to test for `text/html`, not just `html`. That's what I meant.
Just to be consistent with the rest of the source :)